electron/Board.js

```diff
@@ -38,42 +31,120 @@ class Board {
+   * @param runTime {number} is the received running duration of the board (in ms)
+   * @returns {number} the estimated timestamp at which the packet was sent (in ms)
+   */
+  calculateTimestamp(runTime) {
+    if (this.firstRecvTime < 0) {
+      /* TODO: consider using multiple packet offsets to reduce likelihood of noise causing the first receive time to
+      *   deviate too significantly */
+      this.firstRecvTime = Date.now()
+    }
+    if (this.firstRecvOffset < 0) {
+      this.firstRecvOffset = runTime
+    }
+    return runTime - this.firstRecvOffset + this.firstRecvTime
+  }
+
+  /**
+   * Parses the raw packet into a data object
+   * @param buf {Buffer} is the buffer that contains the full udp packet content
+   * @returns {Packet|null} packet with parsed data
+   */
+  parseMsgBuf(buf) {
+    // Packet format:
+    // [ ________ | ________ | ________ ________ ________ ________ | ________ ________ | ________ ... ________ ]
+    // [    id    |   len    |              runTime                |       checkSum    |          data         ]
+    // [  u_int8  |  u_int8  |              u_int32                |        u_int16    |     defined in doc    ]
+    const id = buf.readUInt8(0);
+    const len = buf.readUInt8(1);
+
+    const runTime = buf.readUInt32LE(2);
+    const timestamp = this.calculateTimestamp(runTime);
+
+    const checksum = buf.readUInt16LE(6);
+
+    // currently, data comes after the 2 bytes checksum (at offset 2) 2 + 2 = 4
+    const dataOffset = 8;
+
+    const dataBuf = buf.slice(dataOffset, dataOffset + len)
+    const payloadBuf = buf.slice(0, 6)
+    const sumBuf = Buffer.concat([payloadBuf, dataBuf])
+    const expectedChecksum = Packet.fletcher16(sumBuf)
+
+    if (checksum === expectedChecksum) {
+      const values = []
+      const packetDef = INBOUND_PACKET_DEFS[id]
+
+      let offset = dataOffset;
+
+      for (const [_, parser, __] of packetDef[id]) {
+        const [value, byteLen] = parser(dataBuf, offset);
+        values.push(value);
+        offset += byteLen;
+      }
+
+      return new Packet(id, values, timestamp);
+    } else {
+      console.debug(`check sum check failed for packet id: ${id} from board ip: ${this.address}`)
+      return null
+    }
+  }
+
+  /**
+   * Resets the board status watch dog and increases total bytesReceived
+   * @param byteLen {Number} the number of bytes received
+   * @returns {void}
+   */
+  updateRcvRate(byteLen) {
+    this.resetWatchdog();
+    this.bytesRecv += byteLen;
+  }
+
   /**
    * Takes in a packet and returns a state update
-   *
-   * @param {Packet} packet
+   * @param packet {Packet} a fully formed packet
    */
   processPacket(packet) {
-    this.resetWatchdog();
-    this.bytesRecv += packet.length;
-    const def = this.packets[packet.id];
-    if (def === undefined) return;
+    const { id, values } = packet
+    const packetDef = INBOUND_PACKET_DEFS[id];
```

Comment:
Maybe check whether `packetDef` is defined after this line? But also might be redundant with `parseMsgBuf`

Reply:
redundant